src/node_locks.h

```diff
@@ -0,0 +1,137 @@
+#ifndef SRC_NODE_LOCKS_H_
+#define SRC_NODE_LOCKS_H_
+
+#if defined(NODE_WANT_INTERNALS) && NODE_WANT_INTERNALS
+
+#include <deque>
+#include <memory>  // std::unique_ptr
+#include "node.h"
+#include "node_persistent.h"
+#include "env-inl.h"
+#include "v8.h"
+
+namespace node {
+namespace worker {
+
+class Lock {
+ public:
+  enum Mode {
+    kShared,
+    kExclusive,
+  };
+
+  Lock(Environment* env,
+       Lock::Mode mode,
+       const char* name,
+       v8::Local<v8::Value> released_promise,
+       v8::Local<v8::Value> waiting_promise)
+      : mode_(mode), name_(name) {
+    released_promise_.Reset(env->isolate(), released_promise);
+    waiting_promise_.Reset(env->isolate(), waiting_promise);
+  }
+
+  ~Lock() {
+    waiting_promise_.Reset();
+    released_promise_.Reset();
+  }
+
+  inline const char* name() const { return name_; }
+  inline Mode mode() const { return mode_; }
+
+ private:
+  Mode mode_;
+  const char* name_;
+  node::Persistent<v8::Value> waiting_promise_;
+  node::Persistent<v8::Value> released_promise_;
+};
+
+class LockRequest : public v8::Task {
```

Comment:
As discussed offline, there is no benefit to making this a `v8::Task` versus just a regular libuv-backed object as used elsewhere in Core.